stl/src/vector_algorithms.cpp

```diff
@@ -138,6 +152,29 @@ void* __cdecl __std_swap_ranges_trivially_swappable(void* _First1, void* _Last1,
 
 __declspec(noalias) void __cdecl __std_reverse_trivially_swappable_1(void* _First, void* _Last) noexcept {
 #if !defined(_M_ARM64EC)
+    if (_Byte_length(_First, _Last) >= 128 && _bittest(&__isa_enabled, __ISA_AVAILABLE_AVX512)) {
+        const __m512i _Reverse_8x64bit_avx512 = _mm512_set_epi64(1, 0, 3, 2, 5, 4, 7, 6);
+        const __m512i _Reverse_8x8bit_avx512  = _mm512_set_epi8( //
+            0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, //
+            0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, //
+            0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, //
+            0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15);
+        const void* _Stop_at                  = _First;
+        _Advance_bytes(_Stop_at, _Byte_length(_First, _Last) >> 7 << 6);
+        do {
+            _Advance_bytes(_Last, -64);
+            const __m512i _Left               = _mm512_loadu_si512(static_cast<__m512i*>(_First));
+            const __m512i _Right              = _mm512_loadu_si512(static_cast<__m512i*>(_Last));
+            const __m512i _Left_lane_swapped  = _mm512_permutexvar_epi64(_Reverse_8x64bit_avx512, _Left);
+            const __m512i _Right_lane_swapped = _mm512_permutexvar_epi64(_Reverse_8x64bit_avx512, _Right);
+            const __m512i _Left_reversed      = _mm512_shuffle_epi8(_Left_lane_swapped, _Reverse_8x8bit_avx512);

```

Comment:
What is exactly mean by `__ISA_AVAILABLE_AVX512` ?
AVX512 is separated to sets, like `AVX512F`, `AVX512BW`, etc, and `AVX512F` is supposed to be "default" AVX512, but `_mm512_shuffle_epi8` is for `AVX512BW`

Reply:
As mentioned, the implementation assumes `AVX512BW` to be available. When I searched online, I found two resources.
1. [MSVC Docs](https://docs.microsoft.com/en-us/cpp/build/reference/arch-x64?view=msvc-160)
   > The __AVX512F__, __AVX512CD__, __AVX512BW__, __AVX512DQ__ and __AVX512VL__ preprocessor symbols are defined when the /arch:AVX512 compiler option is specified
2. [DevBlog-Post](https://devblogs.microsoft.com/cppblog/avx-512-auto-vectorization-in-msvc/)
    > /arch:AVX512 is the compiler switch to enable AVX-512 support including auto vectorization. With this switch, the auto vectorizer may vectorize a loop using instructions from the F, VL, BW, DQ, and CD extensions in AVX-512.

I guess that is what StephanTLavavej meant with coarse-grained, but yes, to be sure, one may verify that `cpu_disp.c` checks for all `__cpuid` flags required before setting `__ISA_AVAILABLE_AVX512`.

Reply:
I'm looking into the assembly generated from `cpu_disp.c`...
So my concern is that the value `__ISA_AVAILABLE_AVX512`, which is 6, appears to be never assigned to `__isa_available`, and it can be at most 5.

Reply:
@CaseyCarter acknowledged the check of feature sets, in `cpu_disp.c`, on discord.
> cpu_disp.c sets __isa_available to __ISA_AVAILABLE_AVX512 when bits 16,17,28,30, and 31 are set in EBX by CPUID Leaf 7. IIUC that corresponds to support for AVX512F, AVX512CD, AVX512DQ, AVX512BW, and AVX512VL.